src/main/scala/com/datastax/gatling/plugin/utils/CqlPreparedStatementUtil.scala

```diff
@@ -9,127 +9,152 @@ package com.datastax.gatling.plugin.utils
+          bindable.unset(paramName)
+        } else {
+          bindable
         }
-        boundStatement
       case _ =>
-        paramType match {
-          case (VARCHAR | TEXT | ASCII) =>
-            boundStatement.setString(key, asString(gatlingSession, paramName))
+        paramType.getProtocolCode match {
+          case (VARCHAR | ASCII) =>
+            bindable.setString(key, asString(gatlingSession, paramName))
           case INT =>
-            boundStatement.setInt(key, asInteger(gatlingSession, paramName))
+            bindable.setInt(key, asInteger(gatlingSession, paramName))
           case BOOLEAN =>
-            boundStatement.setBool(key, asBoolean(gatlingSession, paramName))
+            bindable.setBoolean(key, asBoolean(gatlingSession, paramName))
           case (UUID | TIMEUUID) =>
-            boundStatement.setUUID(key, asUuid(gatlingSession, paramName))
+            bindable.setUuid(key, asUuid(gatlingSession, paramName))
           case FLOAT =>
-            boundStatement.setFloat(key, asFloat(gatlingSession, paramName))
+            bindable.setFloat(key, asFloat(gatlingSession, paramName))
           case DOUBLE =>
-            boundStatement.setDouble(key, asDouble(gatlingSession, paramName))
+            bindable.setDouble(key, asDouble(gatlingSession, paramName))
           case DECIMAL =>
-            boundStatement.setDecimal(key, asDecimal(gatlingSession, paramName))
+            bindable.setBigDecimal(key, asDecimal(gatlingSession, paramName))
           case INET =>
-            boundStatement.setInet(key, asInet(gatlingSession, paramName))
+            bindable.setInetAddress(key, asInet(gatlingSession, paramName))
           case TIMESTAMP =>
-            boundStatement.setTimestamp(key, asTimestamp(gatlingSession, paramName))
+            bindable.setInstant(key, asInstant(gatlingSession, paramName))
           case COUNTER =>
-            boundStatement.setLong(key, asCounter(gatlingSession, paramName))
+            bindable.setLong(key, asCounter(gatlingSession, paramName))
           case BIGINT =>
-            boundStatement.setLong(key, asBigInt(gatlingSession, paramName))
+            bindable.setLong(key, asBigInt(gatlingSession, paramName))
           case BLOB =>
-            boundStatement.setBytes(key, asByte(gatlingSession, paramName))
+            bindable.setByteBuffer(key, asByte(gatlingSession, paramName))
           case VARINT =>
-            boundStatement.setVarint(key, asVarInt(gatlingSession, paramName))
+            bindable.setBigInteger(key, asVarInt(gatlingSession, paramName))
           case LIST =>
-            boundStatement.setList(key, asList(gatlingSession, paramName))
+            val dataType = bindable.getType(key)
+            dataType match {
+              case l: ListType => {
+                val memberClz = clzFromCodec(bindable, l.getElementType)
+                bindable.setList(key, asList(gatlingSession, paramName, memberClz), memberClz)
```

Comment:
Have you tested this with nested collections? e.g. `List<List<String>>`

`memberClz` would be `Class[List].asInstanceOf[Class[Any]]`, I'm not sure if that selects the right codec.

Also the code is a bit inefficient in that it looks up the codec twice: once in `clzFromCodec`, and once in `setList`.

Reply:
The previous code already failed in the list-of-lists case:

```
diff --git a/src/test/scala/com/datastax/gatling/plugin/utils/CqlPreparedStatementUtilSpec.scala b/src/test/scala/com/datastax/gatling/plugin/utils/CqlPreparedStatementUtilSpec.scala
index d373c2c..118c431 100644
--- a/src/test/scala/com/datastax/gatling/plugin/utils/CqlPreparedStatementUtilSpec.scala
+++ b/src/test/scala/com/datastax/gatling/plugin/utils/CqlPreparedStatementUtilSpec.scala
@@ -52,6 +52,7 @@ class CqlPreparedStatementUtilSpec extends BaseCassandraServerSpec {
     "setJava" -> Set(1).asJava,
 
     "list" -> List(1),
+    "list-of-lists" -> List(List(1)),
     "listJava" -> List(1).asJava,
 
     "map" -> Map(1 -> 1),
@@ -966,6 +967,12 @@ class CqlPreparedStatementUtilSpec extends BaseCassandraServerSpec {
         result.isSet(15) shouldBe true
       }
 
+      it("should bind with a list of lists") {
+        val result = CqlPreparedStatementUtil.bindParamByOrder(defaultGatlingSession, boundStatementKeys, DataType.Name.LIST, "list-of-lists", 15)
+        result shouldBe a[BoundStatement]
+        result.isSet(15) shouldBe true
+      }
+
       it("should bind with a set") {
         val result = CqlPreparedStatementUtil.bindParamByOrder(defaultGatlingSession, boundStatementKeys, DataType.Name.SET, "set", 16)
         result shouldBe a[BoundStatement]
```

```
[info]   - should bind with a list of lists *** FAILED ***                      
[info]     com.datastax.driver.core.exceptions.InvalidTypeException: Invalid type for list<int> element, expecting java.lang.Integer but got class scala.collection.immutable.$col
on$colon                                                                                                                                                                         
[info]     at com.datastax.driver.core.TypeCodec$AbstractCollectionCodec.serialize(TypeCodec.java:1765)
[info]     at com.datastax.driver.core.TypeCodec$AbstractCollectionCodec.serialize(TypeCodec.java:1740)
[info]     at com.datastax.driver.core.AbstractData.setList(AbstractData.java:349)
[info]     at com.datastax.driver.core.BoundStatement.setList(BoundStatement.java:719)
...
```

This result (plus the idea that list of lists are probably not very likely in Gatling use cases) leads me to argue that while this probably could be improved it feels like something we'd want to address in a follow-up effort, not something that should be addressed as part of the driver upgrade.

Reply:
Works for me, I agree it's not high priority.